File: Leetcode/SortVowelsinaString.rs
// https://leetcode.com/problems/sort-vowels-in-a-string

struct Solution {}

impl Solution {
    pub fn sort_vowels(s: String) -> String {
        const VOWELS: [u8; 10] = [b'a', b'e', b'i', b'o', b'u', b'A', b'E', b'I', b'O', b'U'];
        let mut s_vowels = s
            .chars()
            .enumerate()
            .filter_map(|(ind, c)| {
                let c = c as u8;
                if VOWELS.contains(&c) {
                    Some((c, ind))
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        s_vowels.sort();
        s_vowels.reverse();

        s.chars()
            .map(|c| {
                let cu8 = c as u8;
                if VOWELS.contains(&cu8) {
                    s_vowels.pop().unwrap().0 as char
                } else {
                    c
                }
            })
            .collect::<String>()
    }
}

File: Leetcode/SortVowelsinaString.cpp
// https://leetcode.com/problems/sort-vowels-in-a-string

class Solution {
   public:
    string sortVowels(string s) {
        static const vector<char> Vowels = {'a', 'e', 'i', 'o', 'u'};

        vector<size_t> indxs;
        for (auto i = 0; i < s.size(); ++i) {
            if (Vowels.cend() !=
                find(Vowels.cbegin(), Vowels.cend(), tolower(s[i]))) {
                indxs.push_back(i);
            }
        }

        vector<size_t> indxs2 = indxs;
        sort(indxs2.begin(), indxs2.end(),
             [&](const auto& a, const auto& b) { return s[a] < s[b]; });

        string ans = s;
        for (auto i = 0; i < indxs.size(); ++i) {
            ans[indxs[i]] = s[indxs2[i]];
        }

        return ans;
    }
};